### PR TITLE
Create a bootstrap modal layout object

### PR DESCRIPTION
**Pull request by nsandler1**
Introduces bootstrap modals into crispy forms as a layout object.

Supported by Bootstrap v3+

### CODE REVIEW

**Comment by smithdc1** [comment]
Thanks for this, it looks like a nice addition. 

Could I ask you to have a look at updating the docs, and adding some tests?

Here's a [commit](https://github.com/django-crispy-forms/django-crispy-forms/commit/be31a5d6c6dfd688f70b13268eaff0369caa2d70) that shows how we're currently writing tests. The aim is with this style that it makes them easier to test than lots of HTML in Python strings. 

Do let me know if you need any help with these.

**Comment by nsandler1** [comment]
@smithdc1 Of course! I wanted to hold off on updating the docs and writing tests until my pull request got attention. Thanks for the example.

**Comment by nsandler1** [comment]
Modals are only supported by bootstrap 3 and 4 with no [changes between the versions](https://getbootstrap.com/docs/4.0/migration/#modal) that would cause incompatibility. [f1a8382](f1a8382) only tests bootstrap3 modals and the assumption is a bootstrap4 modal would work the same. If a test for bootstrap4 is requested I'm more than willing to add one, though it would be the exact same test as bootstrap3, just with the bootstrap4 decorator.

**Comment by nsandler1** [comment]
@smithdc1 I'm a bit stuck with how to pass the Python 3.7 check. Any ideas?

File: crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_no_kwargs.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+                        <span aria-hidden="true" style="float: left">&times;</span>
+                    </button>
+                </div>
+                <div class="modal-body">
+                    <div class="form-group" id="div_id_first_name">
+                        <label class="control-label requiredField" for="id_first_name">
+                            first name
+                            <span class="asteriskField">*</span>
+                        </label>
+                        <div class="controls">
```
Comment (smithdc1) [inline crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_no_kwargs.html:17]:
```suggestion
                        <div class=" controls">
```

File: crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_no_kwargs.html [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+            <div class="modal-content">
+                <div class="modal-header">
+                    <h5 class="modal-title" id="modal_title_id">Modal Title</h5>
+                    <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+                        <span aria-hidden="true" style="float: left">&times;</span>
+                    </button>
+                </div>
+                <div class="modal-body">
+                    <div class="form-group" id="div_id_first_name">
+                        <label class="control-label requiredField" for="id_first_name">
```
Comment (smithdc1) [inline crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_no_kwargs.html:13]:
```suggestion
                        <label class=" control-label requiredField" for="id_first_name">
```

**Review by smithdc1 — COMMENTED** [review]
Ah yes, the whole space "thing" with Django 2.2. Not long left to support this version. 

Adding a couple of spaces to the tests should fix it. 

This is now top of my list to have a look at, looks great a a first glance!

File: crispy_forms/bootstrap.py [hunk 3]
```diff
@@ -414,3 +414,57 @@ def __init__(self, field, *args, **kwargs):
...
+            - NOTE: "modal-title" is applied by default
+
+    Example::
+
+        Modal(
+            'field1',
+            Div('field2'),
+            css_id="modal-id-ex",
+            css_class="modal-class-ex,
+            title="This is my modal"
```
Comment (smithdc1) [inline crispy_forms/bootstrap.py:440]:
```suggestion
            title="This is my modal",
```

File: crispy_forms/bootstrap.py [hunk 4]
```diff
@@ -414,3 +414,57 @@ def __init__(self, field, *args, **kwargs):
...
+    """
+
+    template = "%s/layout/modal.html"
+
+    def __init__(self, *fields, **kwargs):
+        self.fields = list(fields)
+        self.template = kwargs.pop("template", self.template)
+
+        self.css_id = kwargs.pop("css_id", "modal_id")
+        self.title = kwargs.pop("title", "Modal Title")
```
Comment (smithdc1) [inline crispy_forms/bootstrap.py:451]: Can we test for escaping here? What happens if I add my title as "Tom & Jerry Survey"?

**Review by smithdc1 — COMMENTED** [review]
Thanks again for the PR! 🏅 

I've got a few notes, mostly nits and/or questions about some edge cases. Appreciate your thoughts on these.

File: crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_with_kwargs.html [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+                <div class="modal-body">
+                    <div class="form-group" id="div_id_first_name">
+                        <label class="control-label requiredField" for="id_first_name">
+                            first name
+                            <span class="asteriskField">*</span>
+                        </label>
+                        <div class="controls">
+                        <input class="form-control inputtext textInput textinput" id="id_first_name" maxlength="5" name="first_name" required type="text">
+                    </div>
+                </div>
```
[the remark this answers is not in the document]
Comment (nsandler1) [reply crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_with_kwargs.html:20]: My intention was that buttons would be added as needed. Only reason I added the close button up top is because bootstrap has that built in so there was no extra code needed. I actually thought about implementing a footer but decided against it because I've personally never been a fan of the literal footer. HTML elements that are needed at the bottom can be added to the bottom of the body using css and/or bootstrap.

I've been using this modal on a personal project and finally decided to actually submit a PR to get it merged. Attached are some examples of how I've been using the modal. Everything in modal was also made using crispy forms.

Example 1 of modal with custom buttons

Example 2 of modal with custom buttons

File: crispy_forms/bootstrap.py [hunk 6]
```diff
@@ -414,3 +414,57 @@ def __init__(self, field, *args, **kwargs):
...
+            Div('field2'),
+            css_id="modal-id-ex",
+            css_class="modal-class-ex,
+            title="This is my modal"
+        )
+    """
+
+    template = "%s/layout/modal.html"
+
+    def __init__(self, *fields, **kwargs):
```
[the remark this answers is not in the document]
Comment (nsandler1) [reply crispy_forms/bootstrap.py:446]: I totally agree. I would of done it how you've shown in your example from the start but I was following the signature style of other layout objects in the codebase. I personally like the defaults being defined in the signature because that way it pops up when typing out the function and it's clear what the defaults are.

File: docs/layouts.rst [hunk 7]
```diff
@@ -301,8 +301,30 @@ These ones live under module ``crispy_forms.bootstrap``.
...
    :align: center
 
-.. _`input_size`:
+- **Modal**: ``Modal`` displays it's fields inside a bootstrap modal that can be customized usign kwargs upon initialization. Only supports Bootstrap v3 or higher::
 
+    Layout(
+        BootstrapModal(
+            # email.help_text was set during the initalization of the django form field
+            Field('email', placeholder="Email", wrapper_class="mb-0"),
+            Button(
```
[the remark this answers is not in the document]
Comment (nsandler1) [reply docs/layouts.rst:310]: Gotcha, I left out how to close the modal with buttons because the bootstrap docs explain how to do that
[via attributes](https://getbootstrap.com/docs/4.0/components/modal/#via-data-attributes) and [via javascript](https://getbootstrap.com/docs/4.0/components/modal/#via-javascript). I'll add these links to the docs.

File: crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_with_kwargs.html [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<form method="post">
+    <div id="id_test" class="modal fade test-class" tabindex="-1" role="dialog" aria-labelledby="id_test-label">
```
[the remark this answers is not in the document]
Comment (nsandler1) [reply crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_with_kwargs.html:2]: Yep, good catch!

File: crispy_forms/tests/test_layout_objects.py [hunk 9]
```diff
@@ -575,3 +576,28 @@ def test_non_ascii_chars_in_container_name(self):
         name = "テスト"
         test_container = Container(name, "val1", "val2")
         assert test_container.css_id == name
+
+    @only_bootstrap3
```
[the remark this answers is not in the document]
Comment (nsandler1) [reply crispy_forms/tests/test_layout_objects.py:580]: Sounds good. I'm running into a problem where `parse_form()` appears to be incorrectly parsing bootstrap4 content. In the bootstrap4 tests, I'm using the same forms as in the bootstrap3 tests. When I print the output of `parse_form()`, I get the following (notice the red circle indicating a class attribute with no actual classes listed):
Screen Shot 2022-01-20 at 6 24 21 PM

The form used in that screen shot was `SampleForm()` and here's the code used in the test:
```
form = SampleForm()
form.helper = FormHelper()
form.helper.layout = Layout(Modal(Field("first_name")))
 ```
Funny thing is, I can't just change the result file to match that output because then `parse_expected()` complains that it was expecting a string following the `class` attribute. Is it possible that the `parse_form()` function has an error?

File: crispy_forms/bootstrap.py (hunk 4 above)
Comment (nsandler1) [reply crispy_forms/bootstrap.py]: Does this still apply even after I changed how kwargs are handled?

File: crispy_forms/bootstrap.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (smithdc1) [reply crispy_forms/bootstrap.py]: Agree, this looks great. Thanks.

File: crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_with_kwargs.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (smithdc1) [reply crispy_forms/tests/results/bootstrap3/test_layout_objects/bootstrap_modal_with_kwargs.html:20]: These look great. 😍 

> ultimately passes the form data to AJAX.

Are you using htmx? That seems to be the latest "hotness" for such things?

File: crispy_forms/tests/test_layout_objects.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (smithdc1) [reply crispy_forms/tests/test_layout_objects.py:580]: My fix for this is in #1208. I think then there may be a couple of tricky space issues with Django 2.2 (Is it April yet?)

File: crispy_forms/bootstrap.py (hunk 4 above)
Comment (smithdc1) [reply crispy_forms/bootstrap.py]: So if I pass in a title as `This is my & modal` the output is `This is my &amp; modal`. I think that's fine as folk can pass their title in as SafeText if they want to this. As far as I know that's standard practice with the exception of `help_text`.

**Comment by smithdc1** [comment]
Ok, tests are passing. Is there anything else before I merge?

**Comment by smithdc1** [comment]
@nsandler1  thanks for the contribution!

**Comment by nsandler1** [comment]
@smithdc1 Nope, looks good to me!
